File: smart_money_analyzer.py
#!/usr/bin/env python3
import json, sys, os, subprocess, datetime

TIER1_WALLETS = [
    "6EDaVsS6enYgJ81tmhEkiKFcb4HuzPUVFZeom6PHUqN3",
    "65kmABTfVidnwPzs5bfeyptpM2ewkkNewnJK6QwUSKXE",
    "3jSHyFJjkWnz73niuzRnjsxSEA6MV3kwKu4ZAFEXGN6f",
    "H2KAvWycyqeHhqkMC1f83Pwju3B2MztLYdXVpNMcDs2V",
    "8fsKLLtvKNanL4ginCaiRS6UfeemY11rSf8U8fN1dJw4",
    "7BWy2mtHNdnWLMZejk5E5sfDDSQfXsjzjy9jPzeDyYk5",
    "MeskxyRYaBhYERcNP2zLVTqFVR43TMX4MihyH7AU4XN",
    "FdwVhbBQhY9rwF5yaZwVH2SK6jhtxj67idgVrrGgAmKS",
    "tCPHCKBKPmVvDCPiv3JmTeXXc7ivPpTRfvPZuzTodir",
    "1aC2FgH1tujX87Bv9yMVbda2sPiRDNCsjMJoJjw3n6C",
    "Ab2iXBkhRwmRtsw7G3PRmY56nerSFYrwR9fi9mcDGZkX",
    "BkqxrgpoWQPc5gx7q225rmoNYoCEdojbarb6Z4iZYhVT",
]

TIER2_WALLETS = [
    "CNDXkovjrP5otMJor1hwSpKWpmY1Qq4vgjBAtQS3Z2TY",
    "43QmFc2QPPGyMrSNuPnhvfs8BFW1XVZYFdbwURtWoo9x",
    "3wccdTM5Ty6ZZC8rpc2GrKUHEScHZmr3iTKjRA6sH2rz",
    "9cxLzxjrTeodcbaEU3KCNGE1a4yFZEcdJ7uEXN378S4U",
    "FhaYN5TDH6JGdkAU1q3vvVeWa7Snbz4dAk53g2XMv9go",
    "DphoNsGZMZ1KVDvLitJREiroFeTePL3YogsDKyP28Ztw",
]

def get_wallet_tier(maker):
    if maker in TIER1_WALLETS:
        return "TIER1"
    elif maker in TIER2_WALLETS:
        return "TIER2"
    return "OTHER"

def analyze_token(token_json):
    try:
        data = json.loads(token_json)
        token = data.get('data', data)
        
        name = token.get('name', token.get('symbol', 'UNKNOWN'))
        address = token.get('address', '')
        
        # Price is nested in some responses
        price_raw = token.get('price', 0)
        if isinstance(price_raw, dict):
            price = float(price_raw.get('price', 0) or 0)
            price_1h_high = float(price_raw.get('price_1h', price) or price)
            buys_1h = int(price_raw.get('buys_1h', 0) or 0)
            sells_1h = int(price_raw.get('sells_1h', 0) or 0)
            vol_5m = float(price_raw.get('volume_5m', 0) or 0)
            vol_1h = float(price_raw.get('volume_1h', 0) or 0)
        else:
            price = float(price_raw or 0)
            price_1h_high = float(token.get('price_1h_high', token.get('high_1h', price)) or price)
            buys_1h = int(token.get('buy_1h', token.get('buys_1h', 0)) or 0)
            sells_1h = int(token.get('sell_1h', token.get('sells_1h', 0)) or 0)
            vol_5m = float(token.get('volume_5m', 0) or 0)
            vol_1h = float(token.get('volume_1h', 0) or 0)
        
        mcap = float(token.get('market_cap', 0) or 0)
        if mcap == 0 and price > 0:
            supply = float(token.get('total_supply', 0) or token.get('circulating_supply', 0) or 0)
            if supply > 0:
                mcap = price * supply
        
        liquidity = float(token.get('liquidity', 0) or 0)
        holders = int(token.get('holder_count', token.get('holders', 0)) or 0)
        
        # Bot rate
        bot_rate = float(token.get('bot_degen_rate', token.get('bot_rate', 0)) or 0)
        
        # Age
        creation_time = token.get('creation_timestamp', 0)
        if creation_time:
            age_min = (datetime.datetime.now(datetime.timezone.utc).timestamp() - creation_time) / 60
        else:
            age_min = 999
        
        # Calculations
        total_trades = buys_1h + sells_1h
        buy_ratio = buys_1h / total_trades if total_trades > 0 else 0
        vol_trend = (vol_5m / (vol_1h / 12)) if vol_1h > 0 else 0
        price_vs_high = (price / price_1h_high * 100) if price_1h_high > 0 else 0
        
        result = {
            'name': name,
            'address': address,
            'price': price,
            'mcap': mcap,
            'liquidity': liquidity,
            'age_min': age_min,
            'vol_5m': vol_5m,
            'vol_1h': vol_1h,
            'vol_trend': vol_trend,
            'buys_1h': buys_1h,
            'sells_1h': sells_1h,
            'buy_ratio': buy_ratio,
            'price_vs_high': price_vs_high,
            'holders': holders,
            'bot_rate': bot_rate,
            'filters_passed': 0,
            'filters_total': 7,
            'checks': {}
        }
        
        # Apply filters
        if age_min < 30:
            result['filters_passed'] += 1
            result['checks']['age'] = True
        else:
            result['checks']['age'] = False
            
        if vol_5m > 5000:
            result['filters_passed'] += 1
            result['checks']['volume_5m'] = True
        else:
            result['checks']['volume_5m'] = False
            
        if vol_trend > 0.20:
            result['filters_passed'] += 1
            result['checks']['vol_trend'] = True
        else:
            result['checks']['vol_trend'] = False
            
        if price_vs_high > 70:
            result['filters_passed'] += 1
            result['checks']['price_vs_high'] = True
        else:
            result['checks']['price_vs_high'] = False
            
        if buy_ratio > 0.6:
            result['filters_passed'] += 1
            result['checks']['buy_ratio'] = True
        else:
            result['checks']['buy_ratio'] = False
            
        if holders > 100:
            result['filters_passed'] += 1
            result['checks']['holders'] = True
        else:
            result['checks']['holders'] = False
            
        if bot_rate < 40:
            result['filters_passed'] += 1
            result['checks']['bot_rate'] = True
        else:
            result['checks']['bot_rate'] = False
        
        result['high_conviction'] = result['filters_passed'] == result['filters_total']
        
        return result
    except Exception as e:
        return {'error': str(e)}

def fetch_token_info(address):
    """Fetch token info using gmgn-cli"""
    try:
        cmd = f"gmgn-cli token info --chain sol --address {address} --raw"
        result = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=15)
        if result.returncode == 0:
            return result.stdout
        return None
    except:
        return None

def main():
    # Read smart money data from stdin or file
    if len(sys.argv) > 1:
        with open(sys.argv[1], 'r') as f:
            data = json.load(f)
    else:
        data = json.load(sys.stdin)
    
    txs = data.get('list', [])
    
    # Get unique buy transactions
    buy_txs = []
    seen = set()
    for tx in txs:
        if tx.get('side') == 'buy':
            addr = tx.get('base_address', '')
            if addr and addr not in seen:
                seen.add(addr)
                buy_txs.append(tx)
    
    print(f"📊 SMART MONEY TRACKER - {datetime.datetime.now(datetime.timezone.utc).strftime('%Y-%m-%d %H:%M UTC')}")
    print(f"Found {len(buy_txs)} unique buy tokens from smart money wallets\n")
    
    # Track results
    high_conviction = []
    filtered_out = {
        'too_old': 0,
        'low_volume': 0,
        'volume_collapsed': 0,
        'dumping': 0,
        'distribution': 0,
        'low_holders': 0,
        'high_bot': 0,
        'other': 0
    }
    
    # Analyze each token
    for tx in buy_txs[:15]:  # Limit to first 15 to avoid rate limits
        addr = tx.get('base_address', '')
        symbol = tx.get('base_token', {}).get('symbol', 'UNKNOWN')
        amount_usd = tx.get('amount_usd', 0)
        maker = tx.get('maker', '')
        is_open = tx.get('is_open_or_close', 0)
        
        # Skip WSOL
        if addr == 'So11111111111111111111111111111111111111112':
            continue
            
        print(f"🔍 Analyzing {symbol} ({addr[:8]}...)...", end=' ', flush=True)
        
        # Fetch token info
        token_json = fetch_token_info(addr)
        if not token_json:
            print("❌ API failed")
            continue
        
        # Save for debugging
        with open(f'/tmp/token_{addr[:8]}.json', 'w') as f:
            f.write(token_json)
        
        result = analyze_token(token_json)
        
        if 'error' in result:
            print(f"❌ Error: {result['error']}")
            continue
        
        passed = result['filters_passed']
        total = result['filters_total']
        
        if result['high_conviction']:
            print(f"✅ ALL FILTERS PASSED")
            
            # Check for cluster signal
            cluster = sum(1 for t in buy_txs if t.get('base_address') == addr) > 1
            
            high_conviction.append({
                'tx': tx,
                'result': result,
                'cluster': cluster,
                'tier': get_wallet_tier(maker)
            })
        else:
            # Track why it failed
            checks = result['checks']
            if not checks.get('age', True):
                filtered_out['too_old'] += 1
                print(f"❌ Too old ({result['age_min']:.0f} min)")
            elif not checks.get('volume_5m', True):
                filtered_out['low_volume'] += 1
                print(f"❌ Low volume (${result['vol_5m']:.0f})")
            elif not checks.get('vol_trend', True):
                filtered_out['volume_collapsed'] += 1
                print(f"❌ Vol collapsed ({result['vol_trend']:.1%})")
            elif not checks.get('price_vs_high', True):
                filtered_out['dumping'] += 1
                print(f"❌ Dumping ({result['price_vs_high']:.1f}%)")
            elif not checks.get('buy_ratio', True):
                filtered_out['distribution'] += 1
                print(f"❌ Distribution ({result['buy_ratio']:.1%})")
            elif not checks.get('holders', True):
                filtered_out['low_holders'] += 1
                print(f"❌ Low holders ({result['holders']})")
            elif not checks.get('bot_rate', True):
                filtered_out['high_bot'] += 1
                print(f"❌ High bot ({result['bot_rate']:.1f}%)")
            else:
                filtered_out['other'] += 1
                print(f"❌ Failed {passed}/{total}")
    
    # Report high conviction signals
    print(f"\n{'='*60}")
    print(f"🔥 HIGH-CONVICTION SIGNALS: {len(high_conviction)}")
    print(f"{'='*60}\n")
    
    for signal in high_conviction:
        tx = signal['tx']
        result = signal['result']
        maker = tx.get('maker', '')
        tier = signal['tier']
        
        print(f"🔥 HIGH-CONVICTION SIGNAL (7/7 filters passed)")
        print(f"[{datetime.datetime.fromtimestamp(tx.get('timestamp', 0), tz=datetime.timezone.utc).strftime('%H:%M')}] {maker[:12]}... ({tier}) — BUY {result['name']} ${tx.get('amount_usd', 0):.2f} [{'FULL' if tx.get('is_open_or_close') else 'PARTIAL'}]")
        print(f"📍 {tx.get('base_address', '')}")
        print(f"💰 MCAP: ${result['mcap']:,.0f} | Price: ${result['price']:.10f} | vs 1h High: {result['price_vs_high']:.1f}%")
        print(f"⏱️ Age: {result['age_min']:.0f} min | Vol 5m: ${result['vol_5m']:,.0f} | Vol Trend: {result['vol_trend']:.1%}")
        print(f"📊 Buy Ratio: {result['buy_ratio']:.1%} | Holders: {result['holders']} | Bot Rate: {result['bot_rate']:.1f}%")
        print(f"🔗 DexScreener: https://dexscreener.com/solana/{tx.get('base_address', '')}")
        print(f"💵 Buy: /buy {tx.get('base_address', '')} 0.1")
        print()
        print("⚡ CONVICTION INDICATORS:")
        print(f"• Multiple wallets: {'YES' if signal['cluster'] else 'NO'}")
        print(f"• Large amount: {'YES' if tx.get('amount_usd', 0) > 500 else 'NO'}")
        print(f"• Full position: {'YES' if tx.get('is_open_or_close') else 'NO'}")
        print(f"• Fresh + high volume: {'YES' if result['age_min'] < 15 and result['vol_5m'] > 50000 else 'NO'}")
        print()
    
    # Report filtered out summary
    total_filtered = sum(filtered_out.values())
    print(f"📊 FILTERED OUT: {total_filtered} low-conviction buys skipped")
    if filtered_out['too_old']:
        print(f"• Too old: {filtered_out['too_old']}")
    if filtered_out['low_volume']:
        print(f"• Low volume: {filtered_out['low_volume']}")
    if filtered_out['volume_collapsed']:
        print(f"• Volume collapsed: {filtered_out['volume_collapsed']}")
    if filtered_out['dumping']:
        print(f"• Dumping: {filtered_out['dumping']}")
    if filtered_out['distribution']:
        print(f"• Distribution: {filtered_out['distribution']}")
    if filtered_out['low_holders']:
        print(f"• Low holders: {filtered_out['low_holders']}")
    if filtered_out['high_bot']:
        print(f"• High bot rate: {filtered_out['high_bot']}")
    
    print(f"\n{'='*60}")
    print(f"Summary: {len(high_conviction)} high-conviction / {total_filtered} filtered / {len(buy_txs)} total buys")
    print(f"{'='*60}")

if __name__ == '__main__':
    main()

File: smart_money_analyzer.js
#!/usr/bin/env node
/**
 * Smart Money Tracker - Direct Analysis
 * Fetches token info for unique BUY transactions from tracked wallets
 */

const { execSync } = require('child_process');

const CURRENT_TIME = Math.floor(Date.now() / 1000);

// Tracked wallets
const TIER1 = new Set([
  '6EDaVsS6enYgJ81tmhEkiKFcb4HuzPUVFZeom6PHUqN3',
  '65kmABTfVidnwPzs5bfeyptpM2ewkkNewnJK6QwUSKXE',
  '3jSHyFJjkWnz73niuzRnjsxSEA6MV3kwKu4ZAFEXGN6f',
  'H2KAvWycyqeHhqkMC1f83Pwju3B2MztLYdXVpNMcDs2V',
  '8fsKLLtvKNanL4ginCaiRS6UfeemY11rSf8U8fN1dJw4',
  '7BWy2mtHNdnWLMZejk5E5sfDDSQfXsjzjy9jPzeDyYk5',
  'MeskxyRYaBhYERcNP2zLVTqFVR43TMX4MihyH7AU4XN',
  'FdwVhbBQhY9rwF5yaZwVH2SK6jhtxj67idgVrrGgAmKS',
  'tCPHCKBKPmVvDCPiv3JmTeXXc7ivPpTRfvPZuzTodir',
  '1aC2FgH1tujX87Bv9yMVbda2sPiRDNCsjMJoJjw3n6C',
  'Ab2iXBkhRwmRtsw7G3PRmY56nerSFYrwR9fi9mcDGZkX',
  'BkqxrgpoWQPc5gx7q225rmoNYoCEdojbarb6Z4iZYhVT'
]);

const TIER2 = new Set([
  'CNDXkovjrP5otMJor1hwSpKWpmY1Qq4vgjBAtQS3Z2TY',
  '43QmFc2QPPGyMrSNuPnhvfs8BFW1XVZYFdbwURtWoo9x',
  '3wccdTM5Ty6ZZC8rpc2GrKUHEScHZmr3iTKjRA6sH2rz',
  '9cxLzxjrTeodcbaEU3KCNGE1a4yFZEcdJ7uEXN378S4U',
  'FhaYN5TDH6JGdkAU1q3vvVeWa7Snbz4dAk53g2XMv9go',
  'DphoNsGZMZ1KVDvLitJREiroFeTePL3YogsDKyP28Ztw'
]);

const ALL_TRACKED = new Set([...TIER1, ...TIER2]);

// Read JSON from stdin
let data = '';
process.stdin.on('data', chunk => data += chunk);
process.stdin.on('end', async () => {
  try {
    const parsed = JSON.parse(data);
    const txs = parsed.list || [];

    // Get only BUY transactions from tracked wallets, most recent first
    const buys = txs
      .filter(tx => tx.side === 'buy' && ALL_TRACKED.has(tx.maker))
      .sort((a, b) => b.timestamp - a.timestamp);

    // Deduplicate by token address (keep most recent)
    const tokenMap = new Map();
    for (const tx of buys) {
      if (!tokenMap.has(tx.base_address)) {
        tokenMap.set(tx.base_address, tx);
      }
    }

    const uniqueBuys = Array.from(tokenMap.values());
    console.log(`📊 Found ${uniqueBuys.length} unique recent buy transactions from tracked wallets\n`);

    const results = {
      passed: [],
      skipped: { tooOld: 0, lowVolume: 0, volumeCollapsed: 0, dumping: 0, distribution: 0, lowHolders: 0, highBotRate: 0, other: 0 }
    };

    for (const tx of uniqueBuys) {
      const tokenAddr = tx.base_address;
      const symbol = tx.base_token?.symbol || 'UNKNOWN';
      const wallet = tx.maker;
      const amount = tx.amount_usd || 0;
      const isFull = tx.is_open_or_close === 1;
      const tier = TIER1.has(wallet) ? 'TIER1' : 'TIER2';
      const walletName = tx.maker_info?.twitter_name || wallet.substring(0, 8) + '...';
      
      console.log(`🔍 Checking ${symbol} from ${walletName} (${tier}) - $${amount.toFixed(2)}`);

      try {
        const cmd = `gmgn-cli token info --chain sol --address ${tokenAddr} --raw`;
        const output = execSync(cmd, { encoding: 'utf8', timeout: 15000 });
        const info = JSON.parse(output);

        // Extract metrics from correct paths based on actual API response
        const creationTime = info.creation_timestamp || info.dev?.ath_token_info?.creation_timestamp || 0;
        const ageMinutes = creationTime > 0 ? (CURRENT_TIME - creationTime) / 60 : 999;
        
        const price = info.price || {};
        const volume5m = parseFloat(price.volume_5m || 0);
        const volume1h = parseFloat(price.volume_1h || 0);
        const volumeTrend = volume1h > 0 ? (volume5m / (volume1h / 12)) * 100 : 0;
        
        const price1hHigh = info.ath_price || parseFloat(price.price_1h || 0);
        const currentPrice = parseFloat(price.price || 0);
        const priceVsHigh = price1hHigh > 0 ? (currentPrice / price1hHigh) * 100 : 100;
        
        const buys1h = price.buys_1h || 0;
        const sells1h = price.sells_1h || 0;
        const buyRatio = (buys1h + sells1h) > 0 ? buys1h / (buys1h + sells1h) : 0;
        
        const holders = info.holder_count || info.stat?.holder_count || 0;
        const botRate = parseFloat(info.stat?.bot_degen_rate || 0) * 100;
        const mcap = info.migration_market_cap || 0;

        // Apply strict filters
        const filters = {
          age: ageMinutes < 30,
          volume: volume5m >= 5000,
          volumeTrend: volumeTrend >= 20,
          price: priceVsHigh >= 70,
          buyRatio: buyRatio >= 0.6,
          holders: holders >= 100,
          botRate: botRate < 40
        };

        const passCount = Object.values(filters).filter(v => v).length;
        const failReasons = [];
        if (!filters.age) failReasons.push(`tooOld(${ageMinutes.toFixed(1)}min)`);
        if (!filters.volume) failReasons.push(`lowVolume($${volume5m.toFixed(0)})`);
        if (!filters.volumeTrend) failReasons.push(`volCollapsed(${volumeTrend.toFixed(1)}%)`);
        if (!filters.price) failReasons.push(`dumping(${priceVsHigh.toFixed(1)}%)`);
        if (!filters.buyRatio) failReasons.push(`distribution(${(buyRatio*100).toFixed(1)}%)`);
        if (!filters.holders) failReasons.push(`lowHolders(${holders})`);
        if (!filters.botRate) failReasons.push(`highBot(${botRate.toFixed(1)}%)`);

        if (passCount === 7) {
          // Check cluster signal
          const clusterWallets = buys.filter(b => b.base_address === tokenAddr).map(b => b.maker);
          const uniqueCluster = [...new Set(clusterWallets)];
          const isCluster = uniqueCluster.length >= 2;

          results.passed.push({
            wallet, walletName, tier, symbol, tokenAddr, amount, isFull, mcap,
            currentPrice, ageMinutes, volume5m, volumeTrend, priceVsHigh,
            buyRatio, holders, botRate, isCluster, clusterCount: uniqueCluster.length,
            isLarge: amount >= 500, isFreshHighVol: ageMinutes < 15 && volume5m > 10000
          });
          console.log(`  ✅ HIGH-CONVICTION SIGNAL - All 7 filters passed!\n`);
        } else {
          const primaryReason = failReasons[0]?.split('(')[0] || 'other';
          if (results.skipped[primaryReason] !== undefined) {
            results.skipped[primaryReason]++;
          } else {
            results.skipped.other++;
          }
          console.log(`  ❌ FAILED (${passCount}/7): ${failReasons.join(', ')}\n`);
        }
      } catch (e) {
        console.log(`  ⚠️ Error: ${e.message}\n`);
        results.skipped.other++;
      }
    }

    // Generate report
    console.log('='.repeat(60));
    console.log('🔥 SMART MONEY HIGH-CONVICTION SIGNAL REPORT');
    console.log('='.repeat(60));
    console.log(`Time: ${new Date(CURRENT_TIME * 1000).toISOString()}`);
    console.log(`Checked: ${uniqueBuys.length} unique buys from tracked wallets`);
    console.log(`Signals: ${results.passed.length} | Filtered: ${Object.values(results.skipped).reduce((a,b)=>a+b,0)}`);
    console.log('');

    if (results.passed.length > 0) {
      for (const r of results.passed) {
        const timeStr = new Date(CURRENT_TIME * 1000).toISOString().replace('T',' ').substring(0,19);
        console.log(`🔥 HIGH-CONVICTION SIGNAL (7/7 filters passed)`);
        console.log(`[${timeStr}] ${r.walletName} (${r.tier}) — BUY $${r.amount.toFixed(2)} ${r.symbol} [${r.isFull?'FULL':'PARTIAL'}]`);
        console.log(`📍 ${r.tokenAddr}`);
        console.log(`💰 MCAP: $${r.mcap.toLocaleString()} | Price: $${r.currentPrice.toExponential(3)} | vs 1h High: ${r.priceVsHigh.toFixed(1)}%`);
        console.log(`⏱️ Age: ${r.ageMinutes.toFixed(1)} min | Vol 5m: $${r.volume5m.toLocaleString()} | Vol Trend: ${r.volumeTrend.toFixed(1)}%`);
        console.log(`📊 Buy Ratio: ${(r.buyRatio*100).toFixed(1)}% | Holders: ${r.holders} | Bot Rate: ${r.botRate.toFixed(1)}%`);
        console.log(`🔗 DexScreener: https://dexscreener.com/solana/${r.tokenAddr}`);
        console.log(`💵 Buy: /buy ${r.tokenAddr} 0.1`);
        console.log('');
        console.log('⚡ CONVICTION INDICATORS:');
        console.log(`• Multiple wallets: ${r.isCluster ? 'YES ('+r.clusterCount+')' : 'NO'}`);
        console.log(`• Large amount (>$500): ${r.isLarge ? 'YES' : 'NO'}`);
        console.log(`• Full position open: ${r.isFull ? 'YES' : 'NO'}`);
        console.log(`• Fresh + high volume: ${r.isFreshHighVol ? 'YES' : 'NO'}`);
        console.log('—'.repeat(40));
      }
    } else {
      console.log('🚫 No high-conviction signals found this cycle.');
      console.log('All tracked wallet buys failed one or more strict filters.');
    }

    console.log('\n📊 FILTERED OUT SUMMARY:');
    console.log(`• Too old (>30min): ${results.skipped.tooOld}`);
    console.log(`• Low volume (<$5K 5m): ${results.skipped.lowVolume}`);
    console.log(`• Volume collapsed (<20% of 1h): ${results.skipped.volumeCollapsed}`);
    console.log(`• Dumping (<70% of 1h high): ${results.skipped.dumping}`);
    console.log(`• Distribution (buy ratio <60%): ${results.skipped.distribution}`);
    console.log(`• Low holders (<100): ${results.skipped.lowHolders}`);
    console.log(`• High bot rate (>40%): ${results.skipped.highBotRate}`);
    console.log(`• Other/errors: ${results.skipped.other}`);

  } catch (e) {
    console.error('Fatal error:', e.message);
    process.exit(1);
  }
});
